### ISSUE TITLE
Support for mask only input

### ISSUE THREAD

**Issue by FrankWJW** [facebookresearch/segment-anything-2#147]
I went through the notebook and only found instruction for point and bounding box inputs. But according to the paper it seems sam2 also support for mask input. Is the example for using mask only input will be released in future?

**Comment by heyoeyo** [comment]
A simple way to get this working is to use one of the masks output by the model itself as an input (I believe this is how the mask input is meant to be used, based on SAMv1).

Here's a simple example that runs the decoder twice, once to get a mask and then a second time using only that mask as an input prompt:
```python
import numpy as np
from PIL import Image
from sam2.build_sam import build_sam2
from sam2.sam2_image_predictor import SAM2ImagePredictor

# File paths
img_path = "path/to/image.jpg"
model_path = "path/to/model.pth"
model_cfg = "model_config.yaml"

# Load model
sam2_model = build_sam2(model_cfg, model_path)
predictor = SAM2ImagePredictor(sam2_model)

# Encode image
image = Image.open(img_path)
image = np.array(image.convert("RGB"))
predictor.set_image(image)

# Run the model once to generate masks to use as inputs
_, _, ex_logits = predictor.predict(
    point_coords=np.array([[500, 375]]),
    point_labels=np.array([1]),
    multimask_output=True,
)

# Create an example mask input
mask_input = ex_logits[0, :, :]
mask_input = np.expand_dims(mask_input, 0) # Make shape: 1x256x256
print("Mask input shape:", mask_input.shape)
print("Mask min/max", mask_input.min(), mask_input.max())

# Run SAM with mask input only
masks, scores, logits = predictor.predict(mask_input=mask_input, multimask_output=True)
```

The mask input generally gives poor quality results for me (compared to points or boxes). It seems to be there to support training more so than as a useful feature on it's own, but maybe there's some cases where it can help.

**Comment by ronghanghu** [comment]
Hi @FrankWJW, regarding using mask prompts:
* For image segmentation, you can feed a mask scores as `mask_input` like @heyoeyo mentioned above
* For interactive segmentation in videos, SAM 2 also supports using masks as a prompt -- We have an interface `add_new_mask` for it in the video predictor class `SAM2VideoPredictor`

**Comment by OrangeSodahub** [comment]
@ronghanghu Hi, I would like to know that for **image segmentation**, the `mask_input` can only be the mask scores output from sam itself? I need to use my own binary mask input, how to do that? Thanks!

**Comment by jneeven** [comment]
@OrangeSodahub 

I can't guarantee any correctness or efficiency, but putting this in `SAM2ImagePredictor.predict()` seems to do the trick:

```python
            assert isinstance(mask_input, np.ndarray) and mask_input.dtype == bool
            mask_input = torch.tensor(mask_input, dtype=torch.bool).float()
            assert mask_input.dim() == 2  # HxW

            # Use -10/+10 as logits for neg/pos pixels (very close to 0/1 in prob after sigmoid).
            out_scale, out_bias = 20.0, -10.0  # sigmoid(-10.0)=4.5398e-05
            high_res_masks = mask_input * out_scale + out_bias
            mask_input: torch.Tensor = torch.nn.functional.interpolate(
                high_res_masks.expand((1, 1, *high_res_masks.shape)),  # Change from HxW to BxCxHxW
                size=(self.model.image_size // 4, self.model.image_size // 4),
                align_corners=False,
                mode="bilinear",
                antialias=True,  # use antialias for downsampling
            ).squeeze((0))  # Change to CxHxW
```

it's a slight modification of the code in [`SAM2Base._use_mask_as_output()`](https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/sam2/modeling/sam2_base.py#L415).

**Comment by mzhadigerov** [comment]
@ronghanghu What exactly should be the input for `add_new_mask` ? 

```python
            predictor.add_new_mask(
                inference_state=inference_state,
                frame_idx=input_frame_idx,
                obj_id=object_id,
                mask=object_mask,
            )
```

what `object_mask` is? boolean numpy array or what exactly?

**Comment by 12306llp** [comment]
> Hi @FrankWJW, regarding using mask prompts:
> 
> * For image segmentation, you can feed a mask scores as `mask_input` like @heyoeyo mentioned above
> * For interactive segmentation in videos, SAM 2 also supports using masks as a prompt -- We have an interface `add_new_mask` for it in the video predictor class `SAM2VideoPredictor`

This add_new_mask() method seems to directly overlay the input mask at the center of the frame, without recognizing similar patterns in the frame or resizing them. Is my usage incorrect, or do I need to perform other operations before use?

**Comment by yufeixue-ai** [comment]
Hi, I have a related question about prompt formulation. I have downloaded the SA-V validation/test datasets and found that they only contain the original separated frames and their corresponding groundtruth (gt) masks. I am confused about the types of prompts used to predict these masks. To utilize the evaluation provided by the project, we need to formulate the prompts to generate the mask predictions, which are not included in the SA-V validation/test datasets.

I am looking for a solution to overcome the lack of corresponding prompts.
